feat(biome_js_analyze): useIndexOf

vladimir-ivanov opened the pull request:
## Summary
Migrating unicorn rule [prefer-array-index-of](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-array-index-of.md)
to biome's `useIndexOf` rule.

vladimir-ivanov commented on .changeset/strong-cars-smile.md:
```diff
@@ -0,0 +1,5 @@
+---
+"@biomejs/biome": major
```
not sure if it should be major? please confirm

dyc3 replied:
this should be minor

vladimir-ivanov replied:
thanks, updated

vladimir-ivanov commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
````diff
@@ -0,0 +1,235 @@
+    /// ```
+    ///
+    /// ```js
+    /// const index = foo.findLastIndex(x => x == undefined);
````
in my opinion this should be marked invalid, but according to unicorn rule it is a valid one. 
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-array-index-of.md

please confirm if we should remove the undefined check, and this becomes invalid?
e.g. it gets fixed to foo.findLastIndex(undefined)

Conaclos replied:
Have you found some rationals to their choice of excluding this case?

vladimir-ivanov replied:
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/test/shared/simple-array-search-rule-tests.js tests do not explicitly cover that use case, so I am guessing it comes from, will do some more investigation and leave it if it is not a mistake in their docs

vladimir-ivanov replied:
looks like a mistake in their docs, just made a test:
<img width="954" alt="Screenshot 2025-05-24 at 17 06 06" src="https://github.com/user-attachments/assets/f74f4aec-77ba-4991-a9c9-280665fadb6f" />

will remove the check

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,235 @@
+mod types;
+
+use crate::JsRuleAction;
+use crate::lint::nursery::use_index_of::{
+    callback_arrow_function_match::callback_arrow_function_match,
+    callback_function_match::callback_function_match, types::JsSyntaxMatchPair,
+};
```
I think we can place these utilities at the end of the current file.
This is what we usually do when the number of utilities is not so large.

vladimir-ivanov replied:
personally prefer small functions, easy to navigate around, getting lost when I have to scroll more than 200 lines of code myself :-) , what my thinking is - if the code is split like this, it becomes easier to discover/ reuse?
but if you say so I can move them in? should not be a big deal - please confirm

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/use_index_of/callback_arrow_function_match.rs:
```diff
@@ -0,0 +1,25 @@
+use biome_js_syntax::{JsArrowFunctionExpression, JsSyntaxToken};
+
+pub fn callback_arrow_function_match(
+    function: &JsArrowFunctionExpression,
+    member_name_token: &JsSyntaxToken,
+) -> Option<JsSyntaxMatchPair> {
+    if function.async_token().is_some() {
+        return None;
+    }
+
+    let parameters = function.parameters().ok()?;
+    let parameter_name = extract_parameter_name(&parameters)?;
+    let body = function.body().ok()?;
+
+    let matched = find_index_comparable_expression(&body, &parameter_name, false);
+
+    matched.as_ref().map(|token_match| JsSyntaxMatchPair {
+        matching_array_element: token_match.clone(),
+        member_name: member_name_token.clone(),
```
To avoid cloning these nodes later, we could take ownership of them.
This should be possible in combination of the change I proposed with the `match` in `use_index_of.rs`.

```suggestion
    function: JsArrowFunctionExpression,
    member_name_token: JsSyntaxToken,
) -> Option<JsSyntaxMatchPair> {
    if function.async_token().is_some() {
        return None;
    }

    let parameters = function.parameters().ok()?;
    let parameter_name = extract_parameter_name(&parameters)?;
    let body = function.body().ok()?;

    let matched = find_index_comparable_expression(&body, &parameter_name, false);

    matched.as_ref().map(|token_match| JsSyntaxMatchPair {
        matching_array_element: token_match,
        member_name: member_name_token,
```

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/use_index_of/callback_function_match.rs:
```diff
@@ -0,0 +1,38 @@
+    Some(parameters.first().unwrap().unwrap().to_trimmed_string())
+}
+
+pub fn callback_function_match(
+    function: &JsFunctionExpression,
+    member_name_token: &JsSyntaxToken,
+) -> Option<JsSyntaxMatchPair> {
+    if function.async_token().is_some() || function.star_token().is_some() {
+        return None;
+    }
+
+    let function_parameters = function.parameters().unwrap().items();
+    let parameter_name = extract_function_parameter_name(&function_parameters)?;
+    let binding = function.body().ok()?;
+    let body = binding
+        .syntax()
+        .descendants()
+        .find_map(AnyJsFunctionBody::cast)?;
+
+    let matched = find_index_comparable_expression(&body, &parameter_name, true);
+
+    matched.as_ref().map(|token_match| JsSyntaxMatchPair {
+        matching_array_element: token_match.clone(),
+        member_name: member_name_token.clone(),
+    })
```
To avoid cloning these nodes later, we could take ownership of them.
This should be possible in combination of the change I proposed with the match in use_index_of.rs.

```suggestion
pub fn callback_function_match(
    function: JsFunctionExpression,
    member_name_token: JsSyntaxToken,
) -> Option<JsSyntaxMatchPair> {
    if function.async_token().is_some() || function.star_token().is_some() {
        return None;
    }

    let function_parameters = function.parameters().unwrap().items();
    let parameter_name = extract_function_parameter_name(&function_parameters)?;
    let binding = function.body().ok()?;
    let body = binding
        .syntax()
        .descendants()
        .find_map(AnyJsFunctionBody::cast)?;

    let matched = find_index_comparable_expression(&body, &parameter_name, true);

    matched.as_ref().map(|token_match| JsSyntaxMatchPair {
        matching_array_element: token_match,
        member_name: member_name_token,
    })
```

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/use_index_of/find_index_comparable_expression.rs:
```diff
@@ -0,0 +1,78 @@
+        return None;
+    }
+
+    let (left, right) = (expression.left().unwrap(), expression.right().unwrap());
```
`unwrap()` should be avoided. Instead, we could just return `None` is something go wrong.

```suggestion
    let (left, right) = (expression.left().unwrap(), expression.right()?);
```

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/use_index_of/find_index_comparable_expression.rs:
```diff
@@ -0,0 +1,78 @@
+    expression: &JsBinaryExpression,
+    parameter_name: &String,
+) -> Option<JsSyntaxNode> {
+    if expression.operator_token().ok()?.text_trimmed() != "===" {
```
I think it is cleaner to compare the token kind:

```suggestion
    if expression.operator_token().ok()?.kind() != T![===] {
```

or

```suggestion
    if expression.operator_token().ok()?.kind() != JsSyntaxKing::EQ3 {
```

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/use_index_of/find_index_comparable_expression.rs:
```diff
@@ -0,0 +1,78 @@
+
+    let (left, right) = (expression.left().unwrap(), expression.right().unwrap());
+
+    let matching_side = if left.to_trimmed_string() == *parameter_name {
+        right
+    } else if right.to_trimmed_string() == *parameter_name {
```
Ideally we would Ike to avoid allocating strings.

vladimir-ivanov replied:
elaborate please?

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/use_index_of/find_index_comparable_expression.rs:
```diff
@@ -0,0 +1,78 @@
+        return None;
+    }
+
+    Some(matching_side.syntax().clone())
```
We could avoid the `clone()`:
```suggestion
    Some(matching_side.into_syntax())
```

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/use_index_of/find_index_comparable_expression.rs:
```diff
@@ -0,0 +1,78 @@
+                || JsLogicalExpression::can_cast(node.kind())
+        })
+        .collect();
+
+    let binary_expressions: Vec<_> = body
+        .syntax()
+        .descendants()
+        .filter_map(JsBinaryExpression::cast)
+        .collect();
+
+    if binary_expressions.len() != 1 || !invalid_expressions.is_empty() {
+        return None;
+    }
```
To return early as soon as possible and avoid extra allocations, I could split the conditionals into two:

```suggestion

    if !invalid_expressions.is_empty() {
        return None;
    }

    let binary_expressions: Vec<_> = body
        .syntax()
        .descendants()
        .filter_map(JsBinaryExpression::cast)
        .collect();

    if binary_expressions.len() != 1 {
        return None;
    }
```

Conaclos reviewed (commented):
Hi and thanks for your contribution! This is already in good shape :) I left some suggestions and comments.

vladimir-ivanov commented:
any idea on why the benchmark is failing please?

vladimir-ivanov commented:
> It seems a temporary issue on GitHub's side, see https://www.githubstatus.com/

thank you, is there any chance to rerun builds without pushing another commit? failing to see an option :-( thanks

siketyan commented:
Dispatched a re-run. Feel free to ask me :)

vladimir-ivanov commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,235 @@
+            return None;
+        }
+
+        let member_name_token = member_expression.syntax().last_token()?;
```
@Conaclos moved all functions in use_index_of.rs, please let me know if you have any ideas on this?

vladimir-ivanov commented:
could anyone review again please? @Conaclos addressed all comments, one is pending four feedback/ suggestion for better code and merged everything in one file for consistency (all fn refs are now inside use_index_of.rs file) thank you

vladimir-ivanov commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,220 @@
+                rule_category!(),
+                node.range(),
+                markup! {
+                    "Prefer `Array#{indexOf,lastIndexOf}()` over `Array#{findIndex,findLastIndex}()` when looking for the index of an item."
```
@arendjr what do you mean please?

vladimir-ivanov replied:
oh, guessing dynamically to read it from cox or _state, sure, let me try, thanks

vladimir-ivanov commented:
@arendjr please do another review when you have a chance? hope I covered everything this time :-).

siketyan commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,363 @@
+
+    let (left, right) = (expression.left().ok()?, expression.right().ok()?);
+
+    let matching_side = if left.syntax().to_string().trim() == *parameter_name {
```
I am on mobile so couldn't check, but there is something like this if I remember correctly:

```suggestion
    let matching_side = if left.syntax().to_text_trimmed() == *parameter_name {
```

vladimir-ivanov commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,363 @@
+    Some(parameters.first().unwrap().unwrap().to_trimmed_string())
+}
+
+pub fn callback_function_match(
```
should not be, forgot to remove after moving the function from dedicated file to here, thanks for flagging, addressed

vladimir-ivanov commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,363 @@
+    })
+}
+
+declare_lint_rule! {
```
of course, addressing

vladimir-ivanov commented:
@siketyan please check if I missed anything? believe addressed all your comments, thanks :-)

ematipico commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,362 @@
+    parameter_name: &String,
+    return_statement_required: bool,
+) -> Option<JsSyntaxNode> {
+    let has_invalid_expression = body.syntax().descendants().find(|node| {
+        JsAssignmentExpression::can_cast(node.kind())
+            || JsVariableDeclaration::can_cast(node.kind())
+            || JsLogicalExpression::can_cast(node.kind())
+    });
+
+    if has_invalid_expression.is_some() {
+        return None;
+    }
+
+    let mut binary_expressions = body
+        .syntax()
+        .descendants()
+        .filter_map(JsBinaryExpression::cast);
+
+    let binary_expression = binary_expressions.next()?;
+    if binary_expressions.next().is_some() {
+        return None;
+    }
+
+    let mut return_statements = body
+        .syntax()
+        .descendants()
+        .filter_map(JsReturnStatement::cast);
```
It seems a bit wasteful to use three iterators to loop over the same thing. Wouldn't it be more efficient to use just one loop and collect the information we need?

vladimir-ivanov replied:
valid point, will try to merge all  into one with filters, thanks

ematipico commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,362 @@
+        return None;
+    }
+
+    Some(parameters.first().unwrap().unwrap().to_trimmed_string())
```
Let's not use unwrap because it panics. Let's use a safer approach

vladimir-ivanov replied:
removed unwrap in favour of ? and ok() - is that any better please?

ematipico commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,362 @@
+        return None;
+    }
+
+    let function_parameters = function.parameters().unwrap().items();
```
Let's avoid unwrap and use a safer approach

ematipico commented on crates/biome_js_analyze/src/lint/nursery/use_index_of.rs:
```diff
@@ -0,0 +1,362 @@
+            .items()
+            .first()?
+            .ok()
+            .map(|item| item.to_trimmed_string()),
```
There are a lot of `to_trimmed_string` functions. They allocate a new string, and sometimes we use it just for comparison, which is a bit wasteful. Can we use a better approach?

vladimir-ivanov replied:
sure, please let me know what a better approach is and will address

ematipico reviewed (changes requested):
We should remove the use of `unwrap`, and possibly remove the use of `to_trimmed_string` where possible. The latter isn't a blocker because it could be tricky, but the former needs to be addressed

arendjr reviewed (approved):
I think we're there :)

vladimir-ivanov commented:
> I think we're there :)

thanks everyone for reviewing, much appreciated, learned a lot :-)

dyc3 commented on .changeset/strong-cars-smile.md:
```diff
@@ -0,0 +1,4 @@
+---
+"@biomejs/biome": minor
+---
+Added the new rule [`useIndexOf`](https://biomejs.dev/linter/rules/use-index-of), based on the unicorn rule [prefer-array-index-of](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-array-index-of.md)
```
Just one more thing: the changeset needs to include examples of valid and invalid code. A single, minimal sample for each will suffice.

vladimir-ivanov replied:
ok, example of what you expect please? 
personally I think is fine as is / has a link to unicorn page which has details? @dyc3

dyc3 reviewed (approved):
Great work!

vladimir-ivanov commented:
if anyone has the super powers to merge, please do so :-), unsure on the rules of how many admins/ moderators must approve etc. thanks
